# ข้อ 1
> สร้าง function ชื่อ smartDict ซึ่งทำหน้าที่:
- function รับ input เป็น list 2 ตัวที่มีขนาดเท่ากัน โดย list แรกเป็น list ของตัวเลข list ที่สองเป็น list ของ text
- ถ้าค่าใน list แรก เป็นเลขคี่ ให้ค่า dict ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 1 และ value จาก list ตัวที่ 2
- ถ้าค่าใน list แรก เป็นเลขคู่ ให้ค่าใน dict  ที่เป็น output มี key เป็นค่าจาก list ตัวที่ 2 และ value จาก list ตัวที่ 1
- output ของ function คือ dict ที่สร้างขึ้น
###### เช่น
            output1 = smartDict([1,2,3],['apple','banana','orange'])
            จะได้ output1 เป็น {1:'apple','banana':2,3:'orange'}
            
            output2 = smartDict([7,1,4,6],['A','B','E','Zzz'])
            จะได้ output2 เป็น {7:'A', 1:'B', 'E': 4, 'Zzz': 6}

In [214]:
def smartDict (listnumber, listtext, logicV):
      if logicV: 
            retdict = {}
            for index in range(len(listnumber)):
                 retdict[listnumber[index]] = listtext[index]
                 
      else:
            retdict = {}
            for index in range(len(listtext)):
                 retdict[listtext[index]] = listnumber[index]
                    
      return retdict

In [215]:
smartDict([1,2,3],['apple','banana','orange'],True)

{1: 'apple', 2: 'banana', 3: 'orange'}

In [216]:
smartDict([7,1,4,6],['A','B','E','Zzz'],False )

{'A': 7, 'B': 1, 'E': 4, 'Zzz': 6}

## ข้อ 2 
> จากไฟล์ Wuhan_nCov_GNome.txt จงเขียนโปรแกรมเพื่อตอบคำถามต่อไปนี้
- กรดอะมิโนใด (A,T,G,C) มีปริมาณมากสุดและมีเท่าไหร่
- จงหา sequence ที่มีขนาด 3 และ 4 ที่เกิดซ้ำมากที่สุด

In [217]:
import pandas as pd

In [218]:
DNAseq = pd.read_csv ("C:\\Users\\HARUKA\\Downloads\\Wuhan_nCoV_GNome.txt",header = None)
DNAseq.head()

0
0  ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGA...
1  CGAACTTTAAAATCTGTGTGGCTGTCACTCGGCTGCATGCTTAGTG...
2  TAATTACTGTCGTTGACAGGACACGAGTAACTCGTCTATCTTCTGC...
3  TTGCAGCCGATCATCAGCACATCTAGGTTTCGTCCGGGTGTGACCG...
4  CCTGGTTTCAACGAGAAAACACACGTCCAACTCAGTTTGCCTGTTT...

##### กรดอะมิโนใด (A,T,G,C) มีปริมาณมากสุดและมีเท่าไหร่

In [219]:
def countChar(seq,Character):
      charcount = 0
      for char in seq:
            if char == Character:
               charcount = charcount + 1 
      return charcount

In [220]:
countChar('ATTAAAGGTTTATACCTTCCCAGGTAACAAACCAACCAACTTTCGATCTCTTGTAGATCTGTTCTCTAAA','A')

22

In [221]:
ANum = 0
TNum = 0
GNum = 0
CNum = 0
for row in range (DNAseq.shape[0]):
      ANum += countChar(DNAseq.iloc[row][0],'A') 
      TNum += countChar(DNAseq.iloc[row][0],'T') 
      GNum += countChar(DNAseq.iloc[row][0],'G') 
      CNum += countChar(DNAseq.iloc[row][0],'C') 
        
print(f' There are {ANum} As, {TNum} Ts, {GNum} Gs, {CNum} Cs')

 There are 8954 As, 9594 Ts, 5863 Gs, 5492 Cs


 * Ans = กรด T มากที่สุดที่ปริมาณ 9594

##### sequence ที่มีขนาด 3 และ 4 ที่เกิดซ้ำมากที่สุด

##### ขนาด 3 

In [222]:
def populate3charstring(seq):
      str_list = list()
      for i in range(len(seq)-2):
           str_list.append(seq[i:i+3])
      return str_list

In [223]:
populate3charstring('ATTAAAGGTT')

['ATT', 'TTA', 'TAA', 'AAA', 'AAG', 'AGG', 'GGT', 'GTT']

In [224]:
list_all_3char = list()
for row in range(DNAseq.shape[0]):
      list_all_3char = list_all_3char + populate3charstring(DNAseq.iloc[row][0])

In [225]:
len(list_all_3char)

29047

In [226]:
unique_3char = list(set(list_all_3char))

In [227]:
len(unique_3char)

64

In [228]:
dict_3char = {}
for sub_str in unique_3char:
      num = 0
      for s in list_all_3char:
            if sub_str == s:
                num += 1
      dict_3char[sub_str]   = num 

In [229]:
dict_3char

{'CTC': 276,
 'AAC': 595,
 'GGA': 272,
 'CCT': 332,
 'TCG': 110,
 'CTT': 720,
 'TTA': 853,
 'ACC': 364,
 'TGT': 834,
 'CTG': 483,
 'GGC': 217,
 'TCT': 526,
 'TCC': 203,
 'ACA': 789,
 'TTG': 800,
 'TTT': 984,
 'AGG': 317,
 'TAG': 414,
 'TAA': 688,
 'TAT': 601,
 'GCT': 509,
 'AGT': 493,
 'TGG': 540,
 'CCA': 336,
 'CCG': 70,
 'ATT': 749,
 'GCC': 184,
 'AAG': 564,
 'TGC': 530,
 'CAA': 681,
 'TCA': 536,
 'CAT': 472,
 'AGC': 292,
 'GGT': 442,
 'CAG': 424,
 'GTG': 532,
 'GTC': 265,
 'CGC': 92,
 'ACT': 660,
 'AAA': 899,
 'AGA': 582,
 'CGA': 91,
 'GTA': 458,
 'ATC': 336,
 'GGG': 131,
 'CAC': 447,
 'GAT': 430,
 'GCA': 363,
 'CTA': 543,
 'GTT': 686,
 'ACG': 157,
 'GAC': 328,
 'GCG': 86,
 'ATA': 456,
 'CCC': 113,
 'ATG': 701,
 'TTC': 509,
 'TGA': 602,
 'TAC': 591,
 'CGG': 74,
 'AAT': 735,
 'CGT': 166,
 'GAA': 524,
 'GAG': 290}

In [230]:
dict_3char['TAC']

591

In [231]:
maxVal = -100
maxKey = None
for k in dict_3char.keys():
      if dict_3char[k] > maxVal:
            maxVal = dict_3char[k]
            maxKey = k

In [232]:
maxKey

'TTT'

In [233]:
print(f' max 3-long-sequence is {maxKey} with frequency equals {maxVal}')

 max 3-long-sequence is TTT with frequency equals 984


##### ขนาด 4

In [234]:
def populate4charstring(seq):
      str_list = list()
      for i in range(len(seq)-3):
           str_list.append(seq[i:i+4])
      return str_list

In [235]:
populate4charstring('ATTAAAGGTT')

['ATTA', 'TTAA', 'TAAA', 'AAAG', 'AAGG', 'AGGT', 'GGTT']

In [236]:
list_all_4char = list()
for row in range(DNAseq.shape[0]):
      list_all_4char = list_all_4char + populate4charstring(DNAseq.iloc[row][0])

In [237]:
len(list_all_4char)

28619

In [238]:
unique_4char = list(set(list_all_4char))

In [239]:
len(unique_4char)

256

In [240]:
dict_4char = {}
for sub_str in unique_4char:
      num = 0
      for s in list_all_4char:
            if sub_str == s:
                num += 1
      dict_4char[sub_str]   = num 

In [241]:
dict_4char

{'AACT': 192,
 'CCGC': 19,
 'TTCG': 38,
 'ATCC': 50,
 'GAAC': 91,
 'GAAA': 160,
 'AGTT': 171,
 'CTGC': 105,
 'TCGC': 23,
 'CGAC': 14,
 'TGTA': 189,
 'TCAC': 108,
 'GACC': 53,
 'CCTC': 54,
 'CCTT': 115,
 'AACA': 245,
 'ACGC': 33,
 'TCCG': 15,
 'TTTT': 294,
 'GATA': 77,
 'GCAA': 112,
 'TAGT': 133,
 'GGAG': 60,
 'CTTG': 160,
 'CAAT': 169,
 'ACTC': 94,
 'CAGG': 81,
 'ACGT': 61,
 'CTTA': 183,
 'TATC': 75,
 'GACG': 28,
 'TCTT': 201,
 'CGTT': 45,
 'GGAC': 63,
 'GATC': 58,
 'TGTC': 138,
 'TACT': 199,
 'TTCC': 76,
 'AGGC': 59,
 'CCCA': 35,
 'ATCT': 124,
 'GTTG': 194,
 'ATTT': 220,
 'TTTG': 252,
 'CTTC': 128,
 'GTCA': 96,
 'AGGA': 87,
 'GCCA': 75,
 'GTGT': 181,
 'TGCA': 159,
 'TCCT': 76,
 'CGAG': 11,
 'GCGG': 14,
 'AAAA': 272,
 'CAAA': 206,
 'TTGC': 163,
 'TGAA': 188,
 'ATCG': 21,
 'TCGG': 22,
 'ATGA': 179,
 'GCAT': 77,
 'AGTA': 105,
 'AGCA': 83,
 'GAGG': 62,
 'ACCA': 139,
 'GGCG': 19,
 'TCTC': 74,
 'AGAC': 113,
 'TGGT': 235,
 'CGTC': 23,
 'GACT': 93,
 'CCTG': 63,
 'CCAC': 78,
 'CAGT': 108,
 'TT

In [242]:
dict_4char['AACT']

192

In [243]:
maxVal = -100
maxKey = None
for k in dict_4char.keys():
      if dict_4char[k] > maxVal:
            maxVal = dict_4char[k]
            maxKey = k

In [244]:
maxKey

'TGTT'

In [245]:
print(f' max 4-long-sequence is {maxKey} with frequency equals {maxVal}')

 max 4-long-sequence is TGTT with frequency equals 318


## ข้อ 3 
> จากไฟล์ data.xlsx สร้างตาราง
- ใหม่ที่มี 2 columns โดย column แรกชื่อว่า ภาษาอีสาน และมีค่าใน column เป็นค่าจาก column ภาษาอีสานเดิมแต่เปลี่ยน ค่าที่เป็น '/' เป็น ',' และ column ที่สองชื่อว่า ภาษาไทย และมีค่าใน column เป็นค่าจาก column ภาษาไทยเดิมแต่เปลี่ยน ค่าที่เป็น '/' เป็น ',' 
##### เช่น 
                จาก สวัสดี/สุผู้สุคน/เด้อ/จ้า    สวัสดี/ทุกๆคน/นะคะ
                เป็น สวัสดี,สุผู้สุคน,เด้อ,จ้า    สวัสดี,ทุกๆคน,นะคะ
- สร้างตาราง Dictionary โดยตารางประกอบด้วยสอง columns คือ ภาษาอีสาน และ ภาษาไทย โดยค่าใน column ภาษาอีสาน คือคำในภาษาอีสาน และ column ภาษาไทย คือคำแปลเป็นภาษากลางของภาษาอีสานใน column แรก ( ใช้จากตาราง data.xlsx โดยคำที่อยู่ในลำดับเดียวกันของ columns ภาษาอีสาน และ ภาษาไทย คือคำแปลของกันและกัน ถ้าจำนวนคำของทั้งสองเท่ากัน - ถ้าจำนวนคำระหว่างสอง columns ไม่เท่ากันให้ข้าม record นั้นไป )


In [246]:
import pandas as pd

##### สร้าง 2 columns คือ ภาษาอิสาน และภาษาไทย

In [247]:
L = ['ภาษาอีสาน' , 'ภาษาไทย']
Data = pd.DataFrame.from_records([],columns = L) 
for i in range(32):
    df = pd.read_excel('data.xlsx', sheet_name = i, usecols = 'B:C', skipfooter = 1)
    Data = pd.concat([Data,df], axis = 0, ignore_index = True)

In [248]:
Data.head()

ภาษาอีสาน  \
0                          สวัสดี/สุผู้สุคน/เด้อ/จ้า   
1       กิน/ข้าว/ก่อน/เด้อ/จ้า/ ไผ/สิ/ถ่า/เค้า/ดาวน์   
2  เค้า/ดาว/กะ/ขอ/ให้/เค้า/ดาว/อย่าง/มี/ความ/สุข/...   
3  นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...   
4                     ขอ/ให้/เชิญ/แซ่บ/เด้อ/พี่/น้อง   

                                             ภาษาไทย  
0                              สวัสดี/ทุก ๆ คน/นะ/คะ  
1          กิน/ข้าว/ก่อน/นะ/จ้ะ/ใคร/จะ/รอ/เค้า/ดาวน์  
2  เค้า/ดาวน์/ก็/ขอ/ให้/เค้า/ดาน์/อย่าง/มี/ความ/ส...  
3  นี่/จ้า/อา/หาร/เย็น/เป็น/ลา/ซาน/ย่า/กับ/ผัก/สด...  
4                        ขอ/ให้/เชิญ/อร่อย/นะ/ทุก/คน

##### เปลี่ยนค่าที่เป็น '/' เป็น ','

In [249]:
Data.iloc[0][1]

'สวัสดี/ทุก ๆ คน/นะ/คะ'

In [250]:
Data.iloc[0][0]

'สวัสดี/สุผู้สุคน/เด้อ/จ้า'

In [251]:
Data.iloc[0]['ภาษาอีสาน'].replace("/", ",")

'สวัสดี,สุผู้สุคน,เด้อ,จ้า'

In [252]:
Data.iloc[0]['ภาษาไทย'].replace("/", ",")

'สวัสดี,ทุก ๆ คน,นะ,คะ'

##### สร้างตาราง Dictionary

In [253]:
def splitstr(col1): 
    co1 = col1.split('/')
    return co1

In [254]:
splitstr(df.iloc[0]['ภาษาอีสาน'])

['ทัก',
 'ทาย',
 'กัน',
 'ซัมบายดี',
 'สวัสดี',
 'วัน',
 'อัง',
 'คาร',
 'เนาะ',
 'จ้า',
 'อา',
 'กาศ',
 'มื้อ',
 'นี่',
 'หนาว',
 'พี',
 'น้อง',
 'แล่ว',
 'ก่อ',
 'อ่า',
 'ช่วง',
 'หน่า',
 'หนาว',
 'นี่',
 'จ้า',
 'ทาง',
 'ทาง',
 'เมือง',
 'นอก',
 'กะ',
 'สิ',
 'แบบ',
 'หว่า']

In [255]:
splitstr(df.iloc[1]['ภาษาไทย'])

['อัน',
 'พระ',
 'อา',
 'ทิตย์',
 'มัน',
 'ก็',
 'เริ่ม',
 'แบบ',
 'มืด',
 'แล้ว',
 'นะ',
 'คะ',
 'เขา',
 'ก็',
 'เริ่ม',
 'แบบ',
 'ตก',
 'แล้ว',
 'แบบ',
 'นี้',
 'นะ']

In [256]:
list1 = []
list2 = []
for i in range(len(Data)):
    if len(splitstr(Data.iloc[i]['ภาษาอีสาน'])) == len(splitstr(Data.iloc[i]['ภาษาไทย'])): 
        list1 += splitstr(Data.iloc[i]['ภาษาอีสาน']) 
        list2 += splitstr(Data.iloc[i]['ภาษาไทย'])

In [257]:
len(list1)

30852

In [258]:
len(list2)

30852

In [259]:
def trans(col1,col2):
    str = f'{col1} แปลว่า {col2}'
    return str

In [260]:
newnew = []
for i in range(len(list1)):
    if len(list1[i]) == len(list2[i]):
        newnew += [trans(list1[i],list2[i])]

In [261]:
unnewnew = set(newnew)

In [262]:
unnewnew

{'เคิล แปลว่า เคิล',
 'กัง แปลว่า กัง',
 'พัง แปลว่า พัง',
 'เอา แปลว่า เอา',
 'แก้ว แปลว่า แก้ว',
 'บ่อๆ แปลว่า ไม่ๆ',
 'นอย แปลว่า นึง',
 'ญาติ แปลว่า ญาติ',
 'เบียง แปลว่า เบียง',
 'หยับ แปลว่า ขยับ',
 'ขวด แปลว่า ขวด',
 'สัพเพ แปลว่า สัพเพ',
 'นุ่มๆ แปลว่า เด็กๆ',
 'โป๊ะ แปลว่า โป๊ะ',
 'แดก  แปลว่า ร้า ',
 'เป็น แปลว่า น้อง',
 'แจก แปลว่า แจก',
 'หวัด แปลว่า หวัด',
 'หา แปลว่า หา',
 'จั่ง แปลว่า หน้า',
 'กำลัง แปลว่า กำลัง',
 'บ๊อย แปลว่า บ่อย',
 'อีสานๆ แปลว่า อีสานๆ',
 'นั่ง แปลว่า นั่ง',
 'ทิ้ง แปลว่า ทิ้ง',
 'ซนะ แปลว่า ชนะ',
 'มืด แปลว่า มืด',
 'บด แปลว่า บด',
 'โซ แปลว่า โซ',
 'ซั้น แปลว่า ชั้น',
 'หลับ แปลว่า หลับ',
 'เฮา แปลว่า ฉัน',
 'หน่า แปลว่า หน้า',
 'สาม แปลว่า สาม',
 'นึก แปลว่า นึก',
 'หลาย แปลว่า เยอะ',
 'อะ แปลว่า อะ',
 'ป้อง แปลว่า ป้อง',
 'อนุญาติ แปลว่า อนุญาติ',
 'นี้ แปลว่า นี้',
 'อื้อ แปลว่า อื้อ',
 'ทุกคน แปลว่า ทุกคน',
 'คัก แปลว่า มาก',
 'หา  แปลว่า หา ',
 'รอบ แปลว่า รอบ',
 'แรกๆ แปลว่า แรกๆ',
 'เต้า แปลว่า เต้า',
 'เฉิน แปลว่า เฉิน',
 'อะเล็ก แปลว่า อะ